src/contrib/cluster/Akka.Cluster.Sharding/DDataShard.cs

```diff
@@ -0,0 +1,229 @@
+        // configuration on each node.
+        private const int NrOfKeys = 5;
+
+        private readonly ImmutableArray<ORSetKey<EntryId>> _stateKeys;
+
+        public DDataShard(
+            string typeName,
+            ShardId shardId,
+            Props entityProps,
+            ClusterShardingSettings settings,
+            ExtractEntityId extractEntityId,
+            ExtractShardId extractShardId,
+            object handOffStopMessage,
+            IActorRef replicator,
+            int majorityCap)
+        {
+            TypeName = typeName;
+            ShardId = shardId;
+            EntityProps = entityProps;
+            Settings = settings;
+            ExtractEntityId = extractEntityId;
+            ExtractShardId = extractShardId;
+            HandOffStopMessage = handOffStopMessage;
+            Replicator = replicator;
+            MajorityCap = majorityCap;
+
+            RememberedEntitiesRecoveryStrategy = Settings.TunningParameters.EntityRecoveryStrategy == "constant"
+                ? EntityRecoveryStrategy.ConstantStrategy(
+                    Context.System,
+                    Settings.TunningParameters.EntityRecoveryConstantRateStrategyFrequency,
+                    Settings.TunningParameters.EntityRecoveryConstantRateStrategyNumberOfEntities)
+                : EntityRecoveryStrategy.AllStrategy;
+
+            _readConsistency = new ReadMajority(settings.TunningParameters.WaitingForStateTimeout, majorityCap);
+            _writeConsistency = new WriteMajority(settings.TunningParameters.UpdatingStateTimeout, majorityCap);
+            _stateKeys = Enumerable.Range(0, NrOfKeys).Select(i => new ORSetKey<EntryId>($"shard-{typeName}-{shardId}-{i}")).ToImmutableArray();
+            
+            GetState();
+        }
+
+        public void EntityTerminated(IActorRef tref) => this.BaseEntityTerminated(tref);
+        public void DeliverTo(string id, object message, object payload, IActorRef sender) => this.BaseDeliverTo(id, message, payload, sender);
+
+        protected override bool Receive(object message) => WaitingForState(ImmutableHashSet<int>.Empty)(message);
+
+        private ORSetKey<EntryId> Key(EntryId entityId)
+        {
+            var i = (Math.Abs(entityId.GetHashCode()) % NrOfKeys);
+            return _stateKeys[i];
+        }
+
+        private void GetState()
+        {
+            for (int i = 0; i < NrOfKeys; i++)
+            {
+                Replicator.Tell(Dsl.Get(_stateKeys[i], _readConsistency, i));
+            }
+        }
+        private Receive WaitingForState(ImmutableHashSet<int> gotKeys) => message =>
+        {
+            void ReceiveOne(int i)
+            {
+                var newGotKeys = gotKeys.Add(i);
+                if (newGotKeys.Count == NrOfKeys)
+                    RecoveryCompleted();
+                else
+                    Context.Become(WaitingForState(newGotKeys));
```

Comment:
Just a comment and asking because I don't know the answer, but what's the allocation impact of these delegate calls that rely on argument closures? Just a new delegate and a long-lived reference to the closure itself?